src/lib/TopLevel.hs

```diff
@@ -588,10 +589,17 @@ evalRequiredSpecializations
 evalRequiredSpecializations e = do
   forM_ (freeAtomVarsList e) \v -> do
     lookupAtomName v >>= \case
-      TopFunBound _ (SpecializedTopFun _) ->
+      TopFunBound _ (SpecializedTopFun (AppSpecialization _ _)) ->
         queryImpCache v >>= \case
           Nothing -> compileTopLevelFun v
           Just _ -> return ()
+      -- XXX: we might need to toposort these based on dependencies. Or else
+      -- maybe we can be careful about the order in which we request them in the
+      -- first place (rather than querying free vars).
```

Comment:
Can you elaborate?

Reply:
> I don't think the Ix typeclass structure is guaranteed to not form cycles?

Typeclasses with cycles are a problem for us in any case! I haven't tried, but I bet we can make the compiler go into an infinite loop during synthesis if we supply some adversarial instances. I don't think the problem here is any worse than what we already have.

Reply:
Yeah I think you're right. In that case I would suggest removing the comment, because IMO it is only more confusing than helpful (we've established that when things are a DAG, it should all work out, and when they're not we're screwed in many other ways too)